### ISSUE TITLE
MongoError: Field name duplication not allowed with modifiers

### ISSUE THREAD

**Issue by rantav** [aldeed/meteor-simple-schema#221]
The following code raises this error `MongoError: Field name duplication not allowed with modifiers`

```
@Col = new Mongo.Collection('col')

Col.attachSchema
  'members.name':
    type: String
    optional: true
  'members.worked':
    type: String
    optional: true

if Meteor.isServer
  Meteor.startup ->
    Col.remove({}) # clean
    id = Col.insert({members: {name: 'n1', worked: 'nowhere'}})
    Col.update({_id: id}, {$set: {members: {name: 'n1', worked: ''}}}) # this line raises the error
```

Now, I'm not sure whether the bug is actually in simple-schema or mongo itself.
  
A few observations: 

1. If the last line was rewritten as the following, there would not be an error: `Col.update({_id: id}, {$set: {members: {name: 'n1'}}})` (note the removal of `worked: ''`
2. I understand why (in general) mongo does this, it's documented in this open issue https://jira.mongodb.org/browse/SERVER-1050 however, I'm not sure that it's really supposed to do it in this case. 
3. Stepping through the code I see that while I send simply `{$set: {members: {name: 'n1', worked: ''}}}` it is internally translated by simple-schema into `{$set: {members: {name: 'n1'}}, $unset: {members: {worked: ''}}}` so this `$set` on `members` along with the `$unset` of `members` triggers the error on mongo. I realize that simple-schema isn't trying to set and unset the same field (that's what mongo is worried about) however, I suppose that since it's all that hierarchical, mongo thinks that the field `members` is being both set and unset. It might be a bug in mongo or maybe it's intended, but that's how things are... Perhaps a workaround might be to `{$unset 'members.worked': ''}}` but I'm not sure it'll work in all cases (arrays?) or that it's easy to implement. 

So anyway, I hope my description helps and again, I'm not sure whether that's a bug in mongo itself or in simple-collection, but  - as a user, I was surprised...

BTW, is there a simple way to see what actually gets sent to mongo (e.g. the complete post-processed update document with all the `$set` and `$unset` in it?). So far I've been stepping through the code, but if there's an easier way to do that that would be useful. 

Regarding a workaround - as mentioned, if I simply omit the part with `worked: ''` then I get through, and in my case that's workable, but I wonder if there's a more generic solution that would work with empty string fields. Thinking of it, is this perhaps on purpose in simple-schema that fields with empty values are `$unset`?

**Comment by rantav** [comment]
Hi @fongandrew I have added the following, but it doesn't help, I still get `MongoError: Field name duplication not allowed with modifiers`

```
  'members':
     type: {Object}
     optional: true
```

**Comment by aldeed** [comment]
It's probably a bug in the simple-schema filtering code. The code that removes empty strings doesn't seem to be handling this case correctly. Until this can be fixed, it should work to pass `{removeEmptyStrings: false}` as the third argument to your `update` call.

**Comment by rantav** [comment]
Hi @aldeed good to know about `{removeEmptyStrings: false}`. 
BTW, doesn't that seem like a mongodb bug actually? I mean you're not really trying to `$set` and `$unset` the same field here (even if there is a bug with empty string removals), I would not expect mongodb to fail in this case...

**Comment by evliu** [comment]
Maybe a possible solution would be to move the $unset statement to a new ```collection.update``` call

**Comment by Miges** [comment]
{removeEmptyStrings: false} worked for me when I had this error too.

**Comment by aykutyaman** [comment]
{removeEmptyStings: false} worked for me too.

**Comment by evliu** [comment]
I wish that option worked for me, but empty fields makes it fail ```check()``` when I try to match against the schema. I'll have to manually delete those fields so it doesn't break other things by putting unexpected values (like empty strings) into the db

**Comment by tagrudev** [comment]
This is a life saver I just spent several hours on trying to figure out a duplication problem - 

```
MongoError: Cannot update 'field' and 'field' at the same time [409]
```

on a field of [CustomObject]

**Comment by aldeed** [comment]
This should be fixed in 1.3.2. Thanks for your patience.
